File: src/core/rule-matcher.js
'use strict';

/**
 * @file rule-matcher.js
 * @description YARA-inspired Signature Evaluation Engine.
 * 
 * ROLE IN ARCHITECTURE:
 * It receives normalized (decoded) HTTP payloads and evaluates them against compiled
 * `.shield` DSL rules. It executes the boolean logic tree (and, or, match_in, etc).
 * 
 * DATA FLOW:
 * [engine.js] passes `decodedReq` -> `prepareMatchData()` creates flat strings -> 
 *  `evaluate()` recursively checks AST nodes -> `matchAllRules()` outputs Threat Matches.
 * 
 * CRITICAL FALSE POSITIVE (FP) MITIGATION:
 * The legacy stringifier concatenated everything into `md.full`. This caused massive 
 * False Positives on XSS signatures when evaluating benign data (e.g., a blog post about XSS).
 * The new stringifier flattens JSON arrays properly and rules strongly prefer `resolveTarget`
 * over generic `md.full` matches.
 */

const TARGET_MAP = {
  'request.url': 'url', 'request.path': 'path', 'request.body': 'body',
  'request.query': 'queryString', 'request.headers': 'headerString',
  'request.cookies': 'cookieString', 'request.method': 'method',
  'request.useragent': 'userAgent', 'request.user_agent': 'userAgent',
  'request.ip': 'ip', 'request.raw_url': 'rawUrl', 'request.raw_body': 'rawBody',
  'request.session': 'sessionId', 'request.sessionid': 'sessionId',
  'request.timestamp': 'timestamp', 'request.time': 'timestamp',
  'request.geoip': 'geoipString', 'request.geo': 'geoipString',
  'request.fingerprint': 'fingerprintString', 'request.fp': 'fingerprintString',
  'request.rate': 'rateString',
  'request.content_type': 'contentType', 'request.hostname': 'hostname',
  'request.protocol': 'protocol',
};

/**
 * Transforms nested objects into a flat string representation safely.
 * FP MITIGATION: Resolves crashes where arrays/objects in GraphQL/JSON are fed to Object.entries.
 * @param {any} val - The value to stringify.
 * @returns {string} Safe flattened string.
 */
function _safeStringify(val) {
  if (typeof val === 'string') return val;
  if (Array.isArray(val)) return val.map(_safeStringify).join(',');
  if (val !== null && typeof val === 'object') {
    return Object.entries(val).map(([k, v]) => `${k}=${_safeStringify(v)}`).join('&');
  }
  return String(val || '');
}

/**
 * Normalizes the decoded request payload into mapping strings for RegExp execution.
 * @param {Object} d - Decoded request object from `decoder.js`.
 * @returns {Object} Search-optimized data structure.
 */
function prepareMatchData(d) {
  const data = {
    url: d.url || '', path: d.path || '', body: d.body || '',
    method: d.method || '', userAgent: d.userAgent || '', ip: d.ip || '',
    rawUrl: d.rawUrl || '', rawBody: d.rawBody || '',
    queryString: '', headerString: '', cookieString: '', full: '',
    sessionId: d.sessionId || '',
    timestamp: d.timestamp || Date.now(),
    geoipString: '',
    fingerprintString: '',
    rateString: '',
    contentType: (d.headers && (d.headers['content-type'] || d.headers['Content-Type'])) || '',
    hostname: (d.headers && (d.headers['host'] || d.headers['Host'])) || '',
    protocol: d.protocol || 'http',
  };

  if (d.query) data.queryString = _safeStringify(d.query);
  if (d.headers) data.headerString = _safeStringify(d.headers);
  if (d.cookies) data.cookieString = _safeStringify(d.cookies);
  if (d.geoip) data.geoipString = _safeStringify(d.geoip);
  if (d.fingerprint) data.fingerprintString = _safeStringify(d.fingerprint);
  if (d.rate) data.rateString = _safeStringify(d.rate);

  // DEEP DIVE: 'data.full' is a brute-force search space. 
  // Should only be used by `any_of_them` global fallback rules. Do not map explicit targets to this.
  data.full = [data.url, data.queryString, data.body, data.headerString, data.cookieString, data.geoipString, data.fingerprintString].join('\n');
  return data;
}

/**
 * Executes a RegExp pattern against the target string safely.
 * @param {Object} def - Compiled string definition from rule.
 * @param {string} text - The target text surface.
 * @returns {boolean} True if matched.
 */
function testPattern(def, text) { 
  return def?.compiled && text ? def.compiled.test(text) : false; 
}

/**
 * Resolves the specific DSL target string (e.g. `request.body`) mapping to the internal data structure.
 * @param {string} name - The DSL target variable name.
 * @param {Object} rule - The current executing rule.
 * @param {Object} md - The Match Data surface.
 * @returns {string} The resolved string content.
 */
function resolveTarget(name, rule, md) {
  if (rule.targets?.[name]) { 
    const mapped = TARGET_MAP[rule.targets[name]] || rule.targets[name]; 
    return md[mapped] || ''; 
  }
  return md.full;
}

/**
 * Evaluates the AST condition block recursively.
 * @param {Object} node - Current AST node (`and`, `or`, `match`, etc).
 * @param {Object} rule - The executing rule context.
 * @param {Object} md - The normalized search data surface.
 * @returns {boolean} True if the node evaluates positively.
 */
function evaluate(node, rule, md) {
  if (!node) return true;
  switch (node.type) {
    case 'and': return evaluate(node.left, rule, md) && evaluate(node.right, rule, md);
    case 'or': return evaluate(node.left, rule, md) || evaluate(node.right, rule, md);
    case 'not': return !evaluate(node.expr, rule, md);
    case 'match': return testPattern(rule.strings[node.pattern], md.full);
    case 'match_in': return testPattern(rule.strings[node.pattern], resolveTarget(node.target, rule, md));
    case 'any_of_them': return Object.values(rule.strings).some(d => testPattern(d, md.full));
    case 'all_of_them': return Object.values(rule.strings).every(d => testPattern(d, md.full));
    case 'any_of': return node.vars.some(v => testPattern(rule.strings[v], md.full));
    case 'all_of': return node.vars.every(v => testPattern(rule.strings[v], md.full));
    case 'boolean': return node.value;
    default: return false;
  }
}

/**
 * Compiles parsed rules into an optimized matching context.
 * Uses Target Grouping and Aho-Corasick for O(n) fast-path filtering.
 * @param {Array} rules - Raw parsed rules.
 * @returns {Object} Compiled matcher context.
 */
function compileRules(rules) {
  const AhoCorasick = require('./aho-corasick');
  const context = {
    allRules: rules,
    targetGroups: {
      url: [], body: [], headers: [], any: []
    },
    aho: new AhoCorasick(),
    literalRules: new Set(), // Rules that MUST have a literal match to trigger
  };

  for (const rule of rules) {
    let hasLiteral = false;
    let targetCategory = 'any';

    // 1. Grouping by Target to prevent evaluating body rules on empty bodies
    const targets = Object.values(rule.targets || {});
    if (targets.every(t => t.includes('body'))) targetCategory = 'body';
    else if (targets.every(t => t.includes('url') || t.includes('path') || t.includes('query'))) targetCategory = 'url';
    else if (targets.every(t => t.includes('header') || t.includes('cookie') || t.includes('useragent'))) targetCategory = 'headers';

    context.targetGroups[targetCategory].push(rule);

    // 2. Extract Literal Strings for Aho-Corasick
    // If all patterns in the rule's condition require specific strings,
    // and those strings are literals (not regex), we can mandate a fast-path match.
    for (const [name, def] of Object.entries(rule.strings)) {
      if (def.type === 'string' && def.value.length >= 3) {
        // Add literal to Aho-Corasick. Output is the rule reference.
        context.aho.add(def.nocase ? def.value.toLowerCase() : def.value, rule);
        hasLiteral = true;
      }
    }

    if (hasLiteral) {
      context.literalRules.add(rule);
    }
  }

  context.aho.build();
  return context;
}

/**
 * Iterates across all loaded signatures using Aho-Corasick Fast-Path.
 * Outputs matches acting as 'Signals' for engine.js's Unified Risk Aggregator.
 * @param {Object} ctx - Compiled Matcher Context.
 * @param {Object} decodedReq - Request data.
 * @returns {Array} List of triggered match objects containing severity and context.
 */
function matchCompiledRules(ctx, decodedReq) {
  const md = prepareMatchData(decodedReq);
  const matches = [];

  // Determine active rule groups based on request context
  const activeRules = new Set(ctx.targetGroups.any);
  ctx.targetGroups.url.forEach(r => activeRules.add(r)); // URL is always present
  if (md.body) ctx.targetGroups.body.forEach(r => activeRules.add(r));
  if (md.headerString) ctx.targetGroups.headers.forEach(r => activeRules.add(r));

  // 1. Aho-Corasick Fast Path (O(n) scan over entire payload)
  const fullPayloadLower = md.full.toLowerCase();
  const fastPathHits = new Set(ctx.aho.search(fullPayloadLower));

  for (const rule of activeRules) {
    // Fast-path exclusion: If the rule relies on literals and Aho didn't find ANY, skip AST eval.
    if (ctx.literalRules.has(rule) && !fastPathHits.has(rule)) {
      continue;
    }

    // AST Evaluation (Slow Path)
    if (!evaluate(rule.condition, rule, md)) continue;

    const mp = [];
    for (const [name, def] of Object.entries(rule.strings)) {
      if (testPattern(def, md.full)) { 
        const m = md.full.match(def.compiled); 
        mp.push({ name, matched: m ? m[0] : '(matched)' }); 
      }
    }

    matches.push({
      rule: rule.name, 
      tags: rule.tags, 
      severity: rule.meta.severity || 'medium',
      category: rule.tags[0] || rule.meta.category || 'unknown',
      description: rule.meta.description || '', 
      author: rule.meta.author || 'ShieldWall',
      sourceFile: rule.sourceFile || 'inline', 
      matchedPatterns: mp,
    });
  }
  return matches;
}

module.exports = { prepareMatchData, testPattern, evaluate, compileRules, matchCompiledRules, TARGET_MAP };

File: src/core/RuleMatcher.cpp
#include "RuleMatcher.hpp"
#include <chrono>
#include <sstream>
#include <algorithm>

static std::string safeStringify(const std::unordered_map<std::string, std::vector<std::string>>& map) {
    std::string result;
    static const std::unordered_set<std::string> blocked = { "__proto__", "constructor", "prototype" };
    for (auto const& [k, v] : map) {
        if (blocked.count(k)) continue;
        for (auto const& val : v) {
            if (!result.empty()) result += "&";
            result += k + "=" + val;
        }
    }
    return result;
}

static std::string safeStringify(const std::unordered_map<std::string, std::string>& map) {
    std::string result;
    static const std::unordered_set<std::string> blocked = { "__proto__", "constructor", "prototype" };
    for (auto const& [k, v] : map) {
        if (blocked.count(k)) continue;
        if (!result.empty()) result += "&";
        result += k + "=" + v;
    }
    return result;
}

MatchData RuleMatcher::prepareMatchData(const HttpRequest& d) {
    MatchData md;
    md.url = d.url;
    md.path = d.path;
    md.body = d.body;
    md.method = d.method;
    md.userAgent = d.userAgent;
    md.ip = d.ip;
    md.rawUrl = d.url;
    md.rawBody = d.body;
    md.sessionId = "";

    md.queryString = ::safeStringify(d.query);
    md.headerString = ::safeStringify(d.headers);
    md.cookieString = ::safeStringify(d.cookies);

    md.timestamp = std::to_string(std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::system_clock::now().time_since_epoch()
    ).count());

    auto ctIt = d.headers.find("content-type");
    if (ctIt == d.headers.end()) ctIt = d.headers.find("Content-Type");
    if (ctIt != d.headers.end() && !ctIt->second.empty()) md.contentType = ctIt->second[0];

    auto hostIt = d.headers.find("host");
    if (hostIt == d.headers.end()) hostIt = d.headers.find("Host");
    if (hostIt != d.headers.end() && !hostIt->second.empty()) md.hostname = hostIt->second[0];

    md.protocol = "http";
    md.full = md.url + "\n" + md.queryString + "\n" + md.body + "\n" + md.headerString + "\n" + md.cookieString;
    return md;
}

bool RuleMatcher::testPattern(const RuleString& def, const std::string& text) {
    if (!def.hasCompiled || text.empty()) return false;
    return std::regex_search(text, def.compiled);
}

std::string RuleMatcher::resolveTarget(const std::string& name, const ShieldRule& rule, const MatchData& md) {
    static const std::unordered_map<std::string, std::string> TARGET_MAP = {
        { "request.url", "url" }, { "request.path", "path" }, { "request.body", "body" },
        { "request.query", "queryString" }, { "request.headers", "headerString" },
        { "request.cookies", "cookieString" }, { "request.method", "method" },
        { "request.useragent", "userAgent" }, { "request.user_agent", "userAgent" },
        { "request.ip", "ip" }, { "request.raw_url", "rawUrl" }, { "request.raw_body", "rawBody" },
        { "request.session", "sessionId" }, { "request.sessionid", "sessionId" },
        { "request.timestamp", "timestamp" }, { "request.time", "timestamp" },
        { "request.geoip", "geoipString" }, { "request.geo", "geoipString" },
        { "request.fingerprint", "fingerprintString" }, { "request.fp", "fingerprintString" },
        { "request.rate", "rateString" },
        { "request.content_type", "contentType" }, { "request.hostname", "hostname" },
        { "request.protocol", "protocol" }
    };

    auto it = rule.targets.find(name);
    if (it != rule.targets.end()) {
        std::string targetVar = it->second;
        auto mapIt = TARGET_MAP.find(targetVar);
        std::string mapped = mapIt != TARGET_MAP.end() ? mapIt->second : targetVar;
        
        if (mapped == "url") return md.url;
        if (mapped == "path") return md.path;
        if (mapped == "body") return md.body;
        if (mapped == "queryString") return md.queryString;
        if (mapped == "headerString") return md.headerString;
        if (mapped == "cookieString") return md.cookieString;
        if (mapped == "method") return md.method;
        if (mapped == "userAgent") return md.userAgent;
        if (mapped == "ip") return md.ip;
        if (mapped == "rawUrl") return md.rawUrl;
        if (mapped == "rawBody") return md.rawBody;
        if (mapped == "sessionId") return md.sessionId;
        if (mapped == "geoipString") return md.geoipString;
        if (mapped == "fingerprintString") return md.fingerprintString;
        if (mapped == "rateString") return md.rateString;
        if (mapped == "contentType") return md.contentType;
        if (mapped == "hostname") return md.hostname;
        if (mapped == "protocol") return md.protocol;
        if (mapped == "timestamp") return md.timestamp;
    }
    return md.full;
}

bool RuleMatcher::evaluate(const std::shared_ptr<AstNode>& node, const ShieldRule& rule, const MatchData& md) {
    if (!node) return true;
    if (node->type == "and") {
        return evaluate(node->left, rule, md) && evaluate(node->right, rule, md);
    }
    if (node->type == "or") {
        return evaluate(node->left, rule, md) || evaluate(node->right, rule, md);
    }
    if (node->type == "not") {
        return !evaluate(node->expr, rule, md);
    }
    if (node->type == "match") {
        auto it = rule.strings.find(node->pattern);
        if (it != rule.strings.end()) {
            return testPattern(it->second, md.full);
        }
        return false;
    }
    if (node->type == "match_in") {
        auto it = rule.strings.find(node->pattern);
        if (it != rule.strings.end()) {
            return testPattern(it->second, resolveTarget(node->target, rule, md));
        }
        return false;
    }
    if (node->type == "any_of_them") {
        for (auto const& [name, def] : rule.strings) {
            if (testPattern(def, md.full)) return true;
        }
        return false;
    }
    if (node->type == "all_of_them") {
        for (auto const& [name, def] : rule.strings) {
            if (!testPattern(def, md.full)) return false;
        }
        return true;
    }
    if (node->type == "any_of") {
        for (auto const& v : node->vars) {
            auto it = rule.strings.find(v);
            if (it != rule.strings.end() && testPattern(it->second, md.full)) return true;
        }
        return false;
    }
    if (node->type == "all_of") {
        for (auto const& v : node->vars) {
            auto it = rule.strings.find(v);
            if (it == rule.strings.end() || !testPattern(it->second, md.full)) return false;
        }
        return true;
    }
    if (node->type == "boolean") {
        return node->booleanValue;
    }
    return false;
}

MatcherContext RuleMatcher::compileRules(const std::vector<ShieldRule>& rules) {
    MatcherContext ctx;
    ctx.allRules = rules;

    for (auto const& rule : rules) {
        bool hasLiteral = false;
        std::string targetCategory = "any";
        
        bool bodyOnly = true;
        bool urlOnly = true;
        bool headersOnly = true;
        
        if (rule.targets.empty()) {
            bodyOnly = urlOnly = headersOnly = false;
        } else {
            for (auto const& [var, target] : rule.targets) {
                if (target.find("body") == std::string::npos) bodyOnly = false;
                if (target.find("url") == std::string::npos && 
                    target.find("path") == std::string::npos && 
                    target.find("query") == std::string::npos) urlOnly = false;
                if (target.find("header") == std::string::npos && 
                    target.find("cookie") == std::string::npos && 
                    target.find("useragent") == std::string::npos) headersOnly = false;
            }
        }
        
        if (bodyOnly) targetCategory = "body";
        else if (urlOnly) targetCategory = "url";
        else if (headersOnly) targetCategory = "headers";
        
        if (targetCategory == "body") ctx.targetGroupBody.push_back(rule);
        else if (targetCategory == "url") ctx.targetGroupUrl.push_back(rule);
        else if (targetCategory == "headers") ctx.targetGroupHeaders.push_back(rule);
        else ctx.targetGroupAny.push_back(rule);
        
        for (auto const& [name, def] : rule.strings) {
            if (def.type == "literal" && def.pattern.length() >= 3) {
                std::string kw = def.pattern;
                if (def.nocase) {
                    std::transform(kw.begin(), kw.end(), kw.begin(), ::tolower);
                }
                ctx.aho.add(kw, rule.name);
                hasLiteral = true;
            }
        }
        if (hasLiteral) {
            ctx.literalRules.insert(rule.name);
        }
    }
    ctx.aho.build();
    return ctx;
}

std::vector<WafMatch> RuleMatcher::matchCompiledRules(const MatcherContext& ctx, const HttpRequest& decodedReq) {
    MatchData md = prepareMatchData(decodedReq);
    std::vector<WafMatch> matches;

    std::vector<ShieldRule> activeRules = ctx.targetGroupAny;
    activeRules.insert(activeRules.end(), ctx.targetGroupUrl.begin(), ctx.targetGroupUrl.end());
    if (!md.body.empty()) {
        activeRules.insert(activeRules.end(), ctx.targetGroupBody.begin(), ctx.targetGroupBody.end());
    }
    if (!md.headerString.empty()) {
        activeRules.insert(activeRules.end(), ctx.targetGroupHeaders.begin(), ctx.targetGroupHeaders.end());
    }

    std::string fullPayloadLower = md.full;
    std::transform(fullPayloadLower.begin(), fullPayloadLower.end(), fullPayloadLower.begin(), ::tolower);
    
    std::vector<std::string> hits = ctx.aho.search(fullPayloadLower);
    std::unordered_set<std::string> fastPathHits(hits.begin(), hits.end());

    for (auto const& rule : activeRules) {
        if (ctx.literalRules.count(rule.name) && !fastPathHits.count(rule.name)) {
            continue;
        }

        if (!evaluate(rule.condition, rule, md)) continue;

        WafMatch match;
        match.rule = rule.name;
        match.tags = rule.tags;

        auto sevIt = rule.meta.find("severity");
        match.severity = sevIt != rule.meta.end() ? sevIt->second : "medium";

        auto catIt = rule.meta.find("category");
        match.category = catIt != rule.meta.end() ? catIt->second : (!rule.tags.empty() ? rule.tags[0] : "unknown");

        auto descIt = rule.meta.find("description");
        match.description = descIt != rule.meta.end() ? descIt->second : "";

        auto authIt = rule.meta.find("author");
        match.author = authIt != rule.meta.end() ? authIt->second : "ShieldWall";

        match.sourceFile = rule.sourceFile.empty() ? "inline" : rule.sourceFile;

        for (auto const& [name, def] : rule.strings) {
            if (testPattern(def, md.full)) {
                match.matchedPatterns.push_back({ name, "(matched)" });
            }
        }
        matches.push_back(std::move(match));
    }
    return matches;
}
